embassy-rp/src/clocks.rs

```diff
@@ -189,14 +633,147 @@ unsafe fn configure_pll(p: pac::pll::Pll, refdiv: u32, vco_freq: u32, post_div1:
 
     // Wait for PLL to lock
     p.prim().write(|w| {
-        w.set_postdiv1(post_div1);
-        w.set_postdiv2(post_div2);
+        w.set_postdiv1(config.post_div1);
+        w.set_postdiv2(config.post_div2);
     });
 
     // Turn on post divider
     p.pwr().modify(|w| w.set_postdivpd(false));
 }
 
+pub trait GpinPin {
+    fn gpin_number(&self) -> usize;
+    fn pin_number(&self) -> usize;
+}
+
+macro_rules! impl_gpinpin {
+    ($name:ident, $pin_num:expr, $gpin_num:expr) => {
+        impl GpinPin for crate::peripherals::$name {
+            fn gpin_number(&self) -> usize {
+                $gpin_num
+            }
+            fn pin_number(&self) -> usize {
+                $pin_num
+            }
+        }
+    };
+}
+
+impl_gpinpin!(PIN_20, 20, 0);
+impl_gpinpin!(PIN_22, 22, 1);
+
+pub struct Gpin<'d, T: GpinPin> {
+    gpout: PeripheralRef<'d, T>,
+}
+
+impl<'d, T: GpinPin> Gpin<'d, T> {
+    pub fn new(gpout: impl Peripheral<P = T> + 'd) -> Self {
+        into_ref!(gpout);
+
+        unsafe {
+            let p = pac::IO_BANK0.gpio(gpout.pin_number()).ctrl();
+            p.write(|w| w.set_funcsel(0x08));
+        }
+
+        Self { gpout }
+    }
+}
+
+impl<'d, T: GpinPin> Drop for Gpin<'d, T> {
+    fn drop(&mut self) {
+        unsafe {
+            let p = pac::IO_BANK0.gpio(self.gpout.pin_number()).ctrl();
+            p.write(|w| w.set_funcsel(pac::io::vals::Gpio0ctrlFuncsel::NULL.0));
+        }
+    }
+}
+
+pub trait GpoutPin {
+    fn gpout_number(&self) -> usize;
+    fn pin_number(&self) -> usize;
+}
+
+macro_rules! impl_gpoutpin {
+    ($name:ident, $pin_num:expr, $gpout_num:expr) => {
+        impl GpoutPin for crate::peripherals::$name {
+            fn gpout_number(&self) -> usize {
+                $gpout_num
+            }
+            fn pin_number(&self) -> usize {
+                $pin_num
+            }
+        }
+    };
+}
+
+impl_gpoutpin!(PIN_21, 21, 0);
+impl_gpoutpin!(PIN_23, 23, 1);
+impl_gpoutpin!(PIN_24, 24, 2);
+impl_gpoutpin!(PIN_25, 25, 3);
+
+pub struct Gpout<'d, T: GpoutPin> {
+    gpout: PeripheralRef<'d, T>,
+}
+
+impl<'d, T: GpoutPin> Gpout<'d, T> {
+    pub fn new(gpout: impl Peripheral<P = T> + 'd) -> Self {
+        into_ref!(gpout);
+
+        unsafe {
+            let p = pac::IO_BANK0.gpio(gpout.pin_number()).ctrl();
```

Comment:
if you make `trait GpoutPin: crate::gpio::Pin` then you can remove `pin_number()`, you can access `ctrl` through the `Pin` trait like `gpout.io().ctrl()`.

example:
https://github.com/embassy-rs/embassy/blob/master/embassy-rp/src/spi.rs#L475
https://github.com/embassy-rs/embassy/blob/master/embassy-rp/src/spi.rs#L98

Reply:
Another reason for adding the supertrait is that `GpoutPin` should be sealed. (A trait becomes sealed if you add a supertrait that's already sealed)

Reply:
Same for Gpin btw

Reply:
I thought something like this was probably possible, but I wasn't sure where to look for it. Fixed.